### ISSUE TITLE
chore: change bundle rules to generate one sfx module and several minimal modules

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#3714]
As outlined in #3458, there should be bundles for angular2.js and all of its direct children. The angular2.js module will get special `sfx` versions that include all third-party dependencies, as well as all angular2 modules.

Blocked by #3713 

Description copied from #3458:

These will be the bundles that are exported:

 * angular2.js (includes all other bundles)
 * core.js
 * core-no-render.js (core without render, for non-dom contexts)
 * forms.js
 * http.js
 * router.js
 * web_worker.js
 * zone.js

And these are the extensions with which each bundle will be published to code.angularjs.org and npm. Bundles will be in the `bundles` folder of the npm package and will be published to code.angularjs.org as `code.angularjs.org/<version-number>/<bundle-name>`.

 * .js
 * .d.ts (typings are in separate `typings` folder in npm distribution)
 * .js.map
 * .min.js (Using uglify for now, eventually using angular/ts-minify)
 * .dev.js (runs in dev mode)
 * .dev.js.map
 * _testing.js
 * _testing.js.map
 * _testing.d.ts

The angular2 bundle will also get an sfx version, which will include dependencies like Rx, reflect-metadata, traceur-runtime, and will export a global `ng` object.

 * angular2.sfx.js
 * angular2.sfx.min.js
 * angular2.sfx.js.map
 * angular2.sfx.min.js.map
 * angular2_with_testing.js*
 * angular2_with_testing.js.map

The `_testing` bundles will include testing utilities and mocks for the respective module. The core_testing.js will include the bulk of the testing utilities, leaving other testing bundles to mostly just include mocks.

* Includes angular2-testing to de-dupe dependencies.

TODO
 - [ ] Update bundle build rules
 - [ ] Change ambient declarations to use `ng` namespace _instead_ of prefix, i.e. `ng.http` instead of `ngHttp`.
 - [ ] Add automated tests against bundles (try using example e2e tests)
 - [ ] Add readme to bundles folder on code.angularjs.org and npm dist to explain how to use
 - [ ] Make sure there's an issue to document how to use bundles in docs

**Comment by jeffbcross** [comment]
I discussed with @rkirov and he said it would be beneficial to have an additional sfx bundle that would include angular2 and angular2-test, since an angular2-test sfx bundle would re-register transitive dependencies like `Rx`, which could have unwanted side effects.

**Comment by jeffbcross** [comment]
Update: I changed the `-testing` bundles to be `_testing` to be consistent with file and directory names throughout angular.

**Comment by jeffbcross** [comment]
Update: added `debug.js` and `core-no-render.js` to bundles, and added todo to change ambient context to `ng` namespace instead of prefix, to more explicitly represent the bundle hierarchy.

**Comment by jeffbcross** [comment]
After further conversation with @mhevery, debug should _NOT_ be its own bundle, but should be included with core, and its `ELEMENT_PROBE_BINDINGS` export should be included in the root injector. I need to make sure that merely including these bindings doesn't produce side effects. I also should open a separate issue to design/clarify the runtime behavior of the debug utilities, if one does not already exist.

**Comment by jeffbcross** [comment]
Apparently, including the `ELEMENT_PROBE_BINDINGS` will override the `AppViewListener` binding, which would cause compiler to call methods of the debug app listener when creating and destroying views. 

https://github.com/angular/angular/blob/master/modules/angular2/src/core/compiler/view_manager.ts#L30

I propose, as a first step, just including the `debug` lib inside `core` for now, so the `ELEMENT_PROBE_BINDINGS` can be added to the injector when debugging without having to add a separate bundle/import. I will open a separate issue to discuss a good user experience for triggering the mode dynamically from browser console. @mhevery sound good to you?

**Comment by mhevery** [comment]
@jeffbcross SGTM

**Comment by mhevery** [comment]
I think this is stale and should be closed.

**Comment by pkozlowski-opensource** [comment]
Thee are some differences here as compared to the latest proposal in https://github.com/angular/angular/issues/5223

> I think this is stale and should be closed.

At least we should have only one issue with bundling strategy for beta :-)
